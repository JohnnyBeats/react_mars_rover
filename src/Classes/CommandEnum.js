const CommandEnum = Object.freeze({
    MOVE: "M",
    LEFT: "L",
    RIGHT: "R"
});

export default CommandEnum;