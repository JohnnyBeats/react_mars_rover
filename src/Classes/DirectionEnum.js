const DirectionEnum = Object.freeze({
    NORTH: "N",
    EAST: "E",
    SOUTH: "S",
    WEST: "W"
});

export default DirectionEnum;