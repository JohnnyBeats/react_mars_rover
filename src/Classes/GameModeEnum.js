const GameModeEnum = Object.freeze({
    SINGLE_PLAYER: "SinglePlayer",
    MULTIPLAYER: "Multiplayer"
});

export default GameModeEnum;