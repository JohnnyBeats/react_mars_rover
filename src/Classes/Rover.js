import DirectionEnum from "./DirectionEnum";
import CommandEnum from "./CommandEnum";

export class RoverEngine{
    constructor(coordinates, direction, commands, maxGrid, roverTeam){
        this.coordinates = coordinates;
        this.direction = direction;
        this.maxGrid = maxGrid;
        this.commands = commands;
        this.roverTeam = roverTeam || new RoverTeam();
        this.commandIndex = -1;
        this.commandsComplete = false;
        this.degreesRotation = 0;
        this.isActive = false;

        switch (direction) {
            case DirectionEnum.NORTH:
                this.degreesRotation = 0;
                break;
            case DirectionEnum.EAST:
                this.degreesRotation = 90;
                break;
            case DirectionEnum.SOUTH:
                this.degreesRotation = 180;
                break;
            case DirectionEnum.WEST:
                this.degreesRotation = 270;
                break;
            default:
                this.degreesRotation = 0;
                break;
        }

    }

    fellowRoverParking = (coordinates) => {
        return (this.roverTeam.getFellowRovers(this).filter(rover =>
            rover.coordinates[0] === coordinates[0] && rover.coordinates[1] === coordinates[1]).length > 0
        );
    };

    // Assumption - Grid numbers will never be negative and can't go over grid max size, if command is to cross grid it will ignore and run next command
    processCommand = (command) => {
        const moveCoordinates = [this.coordinates[0], this.coordinates[1]];

        switch (command) {
            case CommandEnum.MOVE:
                if(this.direction === DirectionEnum.NORTH && this.coordinates[1] < this.maxGrid[1]){
                    moveCoordinates[1] = moveCoordinates[1] + 1;
                }else if(this.direction === DirectionEnum.EAST && moveCoordinates[0] < this.maxGrid[0]){
                    moveCoordinates[0] = moveCoordinates[0] + 1;
                }else if(this.direction === DirectionEnum.SOUTH && moveCoordinates[1] > 0){
                    moveCoordinates[1] = moveCoordinates[1] - 1;
                }else if(this.direction === DirectionEnum.WEST && moveCoordinates[0] > 0){
                    moveCoordinates[0] = moveCoordinates[0] - 1;
                }
                break;
            case CommandEnum.LEFT:
                if(this.direction === DirectionEnum.NORTH){
                    this.direction = DirectionEnum.WEST;
                }else if(this.direction === DirectionEnum.EAST){
                    this.direction = DirectionEnum.NORTH;
                }else if(this.direction === DirectionEnum.SOUTH){
                    this.direction = DirectionEnum.EAST;
                }else if(this.direction === DirectionEnum.WEST){
                    this.direction = DirectionEnum.SOUTH;
                }
                this.degreesRotation = this.degreesRotation - 90;
                break;
            case CommandEnum.RIGHT:
                if(this.direction === DirectionEnum.NORTH){
                    this.direction = DirectionEnum.EAST;
                }else if(this.direction === DirectionEnum.EAST){
                    this.direction = DirectionEnum.SOUTH;
                }else if(this.direction === DirectionEnum.SOUTH){
                    this.direction = DirectionEnum.WEST;
                }else if(this.direction === DirectionEnum.WEST){
                    this.direction = DirectionEnum.NORTH;
                }

                this.degreesRotation = this.degreesRotation + 90;

                break;
            default:
                if(this.commandIndex < this.commands.length){
                    this.runNextCommand();
                }else{
                    this.commandsComplete = true;
                    this.isActive = false;
                }
        }

        if(moveCoordinates[0] !== this.coordinates[0] || moveCoordinates[1] !== this.coordinates[1]){
            //Can't be on the same block with another rover
            if(!this.fellowRoverParking(moveCoordinates)){
                this.coordinates = moveCoordinates;
            }
        }

    };

    runNextCommand = () => {
        if(!this.commandsComplete){
            this.isActive = true;
            this.commandIndex++;
            const command = this.commands[this.commandIndex];
            this.processCommand(command);
        }
    };

    getDirection = () => {
        return this.direction;
    };

    getPosition = () => {
        return this.coordinates[0] + " " +  this.coordinates[1] + " " + this.direction;
    };

    checkPosition = (positionCoordinates) => {
        return this.coordinates[0] === positionCoordinates[0] && this.coordinates[1] === positionCoordinates[1];
    };
}

export class RoverEngineMultiplayer extends RoverEngine{
    constructor(coordinates, direction, commands, maxGrid, roverTeam, id, name){
        super(coordinates, direction, commands, maxGrid, roverTeam);
        this.id = id;
        this.name = name;
    }
}

export class RoverTeam{
    constructor(){
        this.rovers = [];
    }

    getFellowRovers = (rover) => {
       return this.rovers.filter(fellow => rover !== fellow)
    }
}

export default {RoverEngine, RoverEngineMultiplayer, RoverTeam};