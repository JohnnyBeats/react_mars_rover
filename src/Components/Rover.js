import React from "react";
import DirectionEnum from "../Classes/DirectionEnum";

class Rover extends React.Component {
    render() {
        function getDirectionClassName(direction){
            const roverClassName = "rover";

            switch (direction) {
                case DirectionEnum.NORTH:
                    return roverClassName + " north";
                case DirectionEnum.EAST:
                    return roverClassName + " east";
                case DirectionEnum.SOUTH:
                    return roverClassName + " south";
                case DirectionEnum.WEST:
                    return roverClassName + " west";
                default:
                    return roverClassName;
            }
        }

        return (
            //Added inline style item here to ensure turning is smooth between directions
            <div data-testid="rover" className={getDirectionClassName(this.props.engine.getDirection()) + (this.props.engine.isActive ? " active" : " inActive")} style={{transform: 'rotate(' + this.props.engine.degreesRotation + 'deg)'}}>

            </div>
        )
    }
}

export default Rover;