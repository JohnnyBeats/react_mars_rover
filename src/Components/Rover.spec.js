import React from 'react';
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Rover from "./Rover";
import {RoverEngine, RoverTeam} from "../Classes/Rover";
import DirectionEnum from "../Classes/DirectionEnum";

describe('The rover component', ()=> {
    const testData = () => {
        const maxGrid = [5,5];
        const roverTeam = new RoverTeam();

        roverTeam.rovers.push(new RoverEngine([1, 2], "N", "M", maxGrid, roverTeam));

        return {
            rover: roverTeam.rovers[0]
        }
    };

    it('renders the rover with direction north', ()=>{
        const {rover} = testData();
        const {container} = render(
            <Rover engine={rover}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders the rover with direction south', ()=>{
        const {rover} = testData();
        rover.direction = DirectionEnum.SOUTH;
        const {container} = render(
            <Rover engine={rover}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders the rover with direction east', ()=>{
        const {rover} = testData();
        rover.direction = DirectionEnum.EAST;
        const {container} = render(
            <Rover engine={rover}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders the rover with direction west', ()=>{
        const {rover} = testData();
        rover.direction = DirectionEnum.WEST;
        const {container} = render(
            <Rover engine={rover}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders the rover as active', ()=>{
        const {rover} = testData();
        rover.runNextCommand();
        const {container} = render(
            <Rover engine={rover}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });
});