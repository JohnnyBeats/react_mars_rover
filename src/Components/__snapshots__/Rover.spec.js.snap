// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The rover component renders the rover as active 1`] = `
<div
  class="rover north active"
  data-testid="rover"
  style="transform: rotate(0deg);"
/>
`;

exports[`The rover component renders the rover with direction east 1`] = `
<div
  class="rover east inActive"
  data-testid="rover"
  style="transform: rotate(0deg);"
/>
`;

exports[`The rover component renders the rover with direction north 1`] = `
<div
  class="rover north inActive"
  data-testid="rover"
  style="transform: rotate(0deg);"
/>
`;

exports[`The rover component renders the rover with direction south 1`] = `
<div
  class="rover south inActive"
  data-testid="rover"
  style="transform: rotate(0deg);"
/>
`;

exports[`The rover component renders the rover with direction west 1`] = `
<div
  class="rover west inActive"
  data-testid="rover"
  style="transform: rotate(0deg);"
/>
`;
